---
title: Summer 2022 RNA Institute Bioinformatics Workshop | Andam Lab
author: "Teddy Garcia-Aroca"
date: "Last update: July 13th, 2022"
output:
  html_document: rmdformats::downcute
pdf_document: default
---

# Description.

In this tutorial, we will be showing examples of how to visualize data, from  
common boxplots and violin plots to phylogenetic trees.

    
## First, create a vector to install the packages needed:
  
```{r}
packages <- c("devtools", "dplyr", "plyr", "ggplot2", "readr", "ggpubr",
              "rcompanion", "tidyverse", "ggsignif", "reshape", "ggtree", "phytools", "gridExtra")
```

## Install packages not yet installed (this method allows to install only new packages)

```{r}
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
```

## Load all packages
```{r results='hide', message=FALSE, warning=FALSE}
invisible(lapply(packages, library, character.only = T))
```

Note: after running this line, you may see several warning messages about the dependencies loaded and potential conflicts. Ignore those warnings for the purposes of this workshop.

# Plotting metadata

First, we will be plotting the data that are of interest to us, regarding the number such as the of strains per host. In the following example, we plot a simple boxplot for the number of strains per host in the [Richardson et al. 2018](../manuscript/Richardson2018_multiHostPathogen.pdf) paper.

## 1. Set the working directory to where the files are located. In my case (after mapping the network drive) /Volumes/bioinformaticslab/TGarciaAroca/Project2/

```{r}
setwd("/Volumes/bioinformaticslab/TGarciaAroca/Project2/RStudio/")
```

## 2. Load the data from the folder metadata

This table represents the supplementary table 1 from the [Richardson et al. 2018](../manuscript/Richardson2018_multiHostPathogen.pdf) manuscript.

```{r}
metadata <- read.csv("/Volumes/bioinformaticslab/TGarciaAroca/Project2/metadata/metadata.csv", sep=",", header=T)
#head(metadata) 
#tail(metadata)
length(metadata$Number)
```

As you see, this file contains 800 rows of metadata.


## Note: 

Not all the genomes were found in databases. Only 571 genomes were found and used in our analyses, so we will have to filter out the metadata for the genomes that were not available and could not be used in our analyses. For that reason.

To remove those datapoints for missing genomes, we will need to load the phylogenetic tree and extract the tip labels to be removed.

```{r}
tree_file <- "/Volumes/bioinformaticslab/TGarciaAroca/Project2/datasets/core_genome/raxml/RAxML_bestTree.aureus_raxml_out"
aureus_tree <- read.tree(tree_file)
length(aureus_tree$tip.label)

```

As you can see, this tree has 571 tips (isolates), so we will extract the metadata for those 571 tips from the metadata table above, which contains 800 rows (800 isolates).

For this, we used the some functions from the dplyr package.

```{r}
metadata.filtered <- metadata[metadata$ERR.number.tips.filtered %in% aureus_tree$tip.label,] #Extract values in metadata based on tip labels on tree
#head(metadata.filtered)
#tail(metadata.filtered)
length(metadata.filtered$ERR.number.tips.filtered)
```  

As you see, now we have 571 rows in the new metadata.filtered object, containing all the information we need for the tips (isolates) in our phylogenetic tree.

## 3. Now, lets extract the values needed for plotting and rename the columns.

If we look at the current column names for this dataframe:

```{r}
print(as_tibble(metadata.filtered), n = 2) # printing the first two rows only
```

We see that the column names are a bit confussing. For now, we will leavem them as they are and only select the two that we need:

-*ERR.number.tips.filtered* = Isolate names matched to the tip names on our phylogenetic tree (done by Teddy)

-*Source* = host

Now, lets plot and example:

```{r}
metadata.for.plotting <- dplyr::select(metadata.filtered, ERR.number.tips.filtered, Source) #extract values needed from filtered table
```

Once we have selected the two columns we need, we rename them with more informative column names:

```{r}
colnames(metadata.for.plotting) <- c("Isolate", "Host") # rename columns
length(metadata.for.plotting$Isolate)
```

As you see, we have 571 isolates total, wich matches what we have in our phylogenetic tree.

## 4. Plot the number of isolates per host

In this example, we plot all isolates per host. A simple count plot with ggplot:

```{r}
isolate.per.host.ggplot <- ggplot(metadata.for.plotting, aes(x=Host)) + # load data into ggplot and define axes
  geom_bar() + #Plot a geometric bar
  ggtitle("Number of isolates per host") + # add a title
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) # Rotate labels
isolate.per.host.ggplot
```

As we see here, this is not an easy plot to read. So now, lets plot by Host_group.

```{r}
metadata.for.plotting <- dplyr::select(metadata.filtered, ERR.number.tips.filtered, Source.group, CC) #extract values needed from filtered table
colnames(metadata.for.plotting) <- c("Isolate", "Host_group", "CC")
print(as_tibble(metadata.for.plotting), n = 2)

isolate.per.host.ggplot <- ggplot(metadata.for.plotting, aes(x=Host_group)) + # load data into ggplot and define axes
  geom_bar() + #Plot a geometric bar
  ggtitle("Number of isolates per host") + # add a title
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) # Rotate labels
isolate.per.host.ggplot
```
This is much better!


*An example of how to extract and plot your data from within these data sets is provided in the next few lines*

# Objective 1: Plotting a phylogenetic tree

At this point, you all of you should have your own phylogenetic tree from your analyses in the HPCC cluster. In this tutorial, I provide examples of how to plot and visualize your phylo-tree and append data for visualization purposes.

## First, lets extract the  relevant metadata from above ("metadata.filtered" object) to use in our phylogenetic trees (using the dplyr package). 

We need to extract the hosts and clonal complexes for plotting. We use the same approach we used above:

```{r}
metadata.for.tree <- dplyr::select(metadata.filtered, ERR.number.tips.filtered, Source, CC) #extract values needed from filtered table
length(metadata.for.tree$ERR.number.tips.filtered)
```

## Now, lets plot:

But first, lets change the column names of our dataframe for R to be able to associate the tips of the tree with our metadata.

```{r}
print(as_tibble(metadata.for.tree), n = 2)
```
As we see, the first column should be called "tips" because those are the names of the tips in our tree file (aureus_tree).

```{r}
length(aureus_tree$tip.label)
```

To change the column names:

```{r}
colnames(metadata.for.tree) <- c("tip.label", "Host", "CC")
print(as_tibble(metadata.for.tree), n = 2)
```

Now we plot with ggplot:

```{r}
gg1 <- ggtree(aureus_tree, layout = "circular", ladderize = T) 
gg1 <- gg1 %<+% metadata.for.tree + xlim(NA, 0.8) +
  geom_tippoint(aes(color = factor(Host)), size=3) + 
  geom_treescale(x=0.08, y=35) #+
gg1 <- gg1 + theme(legend.position="right") + labs(tag="A") + ggtitle("S. aureus 571 genomes") +
  theme(plot.title = element_text(size = 12, hjust = 0.1, face = "bold")) +
  scale_colour_discrete(na.translate = F, name="Host") 
gg1
```

As you see, this tree is hard to read. We will plot it with the "Source.group" column in the metadata. We will do it in the same manner we made the previous figure.

```{r}
metadata.for.tree.2 <- dplyr::select(metadata.filtered, ERR.number.tips.filtered, Source.group, CC) #extract values needed from filtered table
colnames(metadata.for.tree.2) <- c("tip.label", "Host_group", "CC")
print(as_tibble(metadata.for.tree.2), n = 2)
```
For your own data, you will need to obtain a specific sequence cluster or clonal complex (CC) as described by Richardson et a. 2018. 

We plot using this new metadata.

Let's plot 
```{r}
gg2 <- ggtree(aureus_tree, layout = "fan", ladderize = T)
gg2 <- gg2 %<+% metadata.for.tree.2 +
  geom_tippoint(aes(color = factor(Host_group)), size=3)
gg2 <- gg2 + theme(legend.position="right") + labs(tag="B") + ggtitle("S. aureus 571 genomes") +
  theme(plot.title = element_text(size = 12, hjust = 0.1, face = "bold")) +
  scale_colour_discrete(na.translate = F, name="Host groups")
gg2
```

Looks better now, but still somewhat convoluted. We will leave it as is for now and we will plot again with more metadata.

Lets look at the number of clonal complexes, as they are informative:

```{r}
unique(metadata.filtered$CC)
```

We have 76 clonal complexes (CCs) in this dataset. That means we have to add that information to our phylogenetic tree. To do so, we extract each group of IDs in our metadata (ERR.number.tips.filtered) that match the tips in our phylogenetic tree for each CC.


```{r}
#first, group the clonal complexes (ccs) with dplyr
CCs <- metadata.filtered %>% group_split(CC)

split_df<-function(list){
  for (i in 1:length(list)){
    assign(paste0("df",i), list[[i]], envir = .GlobalEnv)
  }
}
split_df(CCs)

length(CCs)
```

As you see, we have created many smaller datasets for each clonal complex (CCs), 75 in total (one missing). Now, we can add these data to our tree if we desire to do so. For the purposes of this workshop, we will not worry about CCs, but CCs represent in many case informative data about the specific strains and the hosts they inhabit.

In most cases, we want a midpoint rooted tree, so let's get ours midpoint rooted before plotting.

```{r}
new_aureus_tree <- midpoint.root(aureus_tree)
```

No, we plot the DNA sequences by host as follows:

```{r}
metadata.for.tree.3 <- metadata.filtered %>% select(c("ERR.number.tips.filtered", "Source.group"))
metadata.for.tree.3 <- aggregate(.~Source.group, metadata.for.tree.3, FUN=paste, collapse=",")
host_group <- lapply(metadata.for.tree.3$ERR.number.tips.filtered, function(x){unlist(strsplit(x,split=","))})
names(host_group) <- metadata.for.tree.3$Source.group

gg3 <- ggtree(new_aureus_tree, layout = 'circular')
groupOTU(gg3, host_group, 'host_group') + aes(color=host_group) +
  theme(legend.position="right") + geom_tiplab(size=1, aes(angle=angle))
```
We can also plot the same tree in a rectangular configuration:

```{r}
gg3 <- ggtree(new_aureus_tree, layout = 'rectangular')
groupOTU(gg3, host_group, 'host_group') + aes(color=host_group) +
  theme(legend.position="right") + geom_tiplab(size=1, aes(angle=angle))
```


Here we see an interesting result that was not highlighted by the authors of the manuscript. Apparently, most internal nodes of this phylogeny are made up from *rodents*, which represents an interesting finding on itself. We will discuss these types of findings and their implications in our small group meetings.


*Note: For this initial part, your assignment will be to highlight the clusters that contain your host species. An example is provided here.*

Example:

Lets say you are working with Sheep, Monkey and Bats, you would collapse all other host groups as follows:

1. Find the nodes where Sheep, Bats, and Monkeys appear in your phylogeny.

```{r}

```



```{r}



geom_cladelab(data = dt, 
              mapping = aes(node = node, label = name, 
                          image = image, color = name), 
              geom = "phylopic", offset = 1, offset.text=0.5)
```




