---
pdf_document: default
author: "Teddy Garcia-Aroca"
date: 'Last update: July 27th, 2022'
output:
  html_document: rmdformats::downcute
  pdf_document: default
title: "Project 2 | Andam Lab | Plotting Markov Jumps"
---

# Plotting results from Markov Jumps analyses.

**Note: this is just an example, since an approximation had to be used in BEAST to obtain these results**

In this tutorial, we will be showing examples of how to visualize results from Markov Jumps analyses from [StarBEAST](https://taming-the-beast.org/tutorials/StarBeast-Tutorial/), which can be used for a wide range of phylogenetic analyses and evolutionary data.

## Load packages and metadata as we did previously:
  
```{r results='hide', message=FALSE, warning=FALSE}
packages <- c("ape", "devtools", "dplyr", "plyr", "ggplot2", "readr", "ggpubr",
              "rcompanion", "tidyverse", "ggsignif", "reshape", "ggtree", "phytools", "gridExtra")

# install packages that haven't been installed
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}

# load packages
invisible(lapply(packages, library, character.only = T))

#set working directory
setwd("/Volumes/bioinformaticslab/TGarciaAroca/Project2/RStudio/")

# Load metadata
metadata <- read.csv("/Volumes/bioinformaticslab/TGarciaAroca/Project2/metadata/metadata.csv", sep=",", header=T)
length(metadata$Number)

#Load tree
tree_file <- "/Volumes/bioinformaticslab/TGarciaAroca/Project2/datasets/core_genome/raxml/RAxML_bestTree.aureus_raxml_out"
aureus_tree <- read.tree(tree_file)
length(aureus_tree$tip.label)

#Filter metadata
metadata.filtered <- metadata[metadata$ERR.number.tips.filtered %in% aureus_tree$tip.label,] #Extract 
length(metadata.filtered$ERR.number.tips.filtered)

```

## Now, let's load results from **StarBeas**t.

The results can be found in the folder called "markov_jumps_example" in the github repository (datasets/core_genome/markov_jumps_example/). These results have been adjusted to fit the tree and metadata columns.

```{r}
#Filter metadata
markov.jumps.ncbi <- read.csv("/Volumes/bioinformaticslab/TGarciaAroca/Project2/datasets/core_genome/AMR/abricate/ncbi/abricate_ncbi_results.summary.mod.csv")
```

### Merge the two datasets (metadata and abricate results)

We merge them as we did in previous examples (**Note: since the column names are different for both datasets, we use by.x, and by.x to specify the column names we want to merge)

```{r}
metadata.filtered.abricate <- merge(metadata.filtered, abricate.ncbi, by.x="ERR.number.tips.filtered", by.y="tip.label")
length(metadata.filtered.abricate$ERR.number.tips.filtered)
```

We see that we only have 565 data points now. That means no AMR genes were found in some genomes (571-565= 6). The number of genomes without any AMR genes is low because multi-drug resistance is common in S. aureus.


## Plot AMR gene counts per host group

```{r}
metadata.filtered.abricate.2 <- metadata.filtered.abricate %>% 
	dplyr::group_by(Source.group)
	
metadata.filtered.abricate.2$Source.group <- with(metadata.filtered.abricate.2, reorder(Source.group, -NUM_FOUND))
	ggplot(metadata.filtered.abricate.2, aes(x =Source.group, y = log(NUM_FOUND), fill=Source.group)) + 
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
   labs(
    title = "Comparisons in number AMR genes per host groups",
    x = "Host groups",
    y = "Log(Number of AMR genes)"
    )
```

## Make a boxplot instead

Since this is numericald data, we can make a boxplot to look at the distribution of values

```{r}
metadata.filtered.abricate.2 <- metadata.filtered.abricate %>% 
	dplyr::group_by(Source.group)
	
metadata.filtered.abricate.2$Source.group <- with(metadata.filtered.abricate.2, reorder(Source.group, -NUM_FOUND))
	ggplot(metadata.filtered.abricate.2, aes(x =Source.group, y = NUM_FOUND, fill=Source.group)) + 
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
   labs(
    title = "Comparisons in number AMR genes per host groups",
    x = "Host groups",
    y = "Number of AMR genes"
    )
```

As you see, this is way more informative. Surprisingly, more AMR genes were found in pigs (another result not presented in the manuscript). We see that there is some variation across all the host species and some outliers. We can do some statistics to see if those results are significantly different (see below).

## Plot violin plots

First we create a vector of the **pairwise comparisons** we want to make:

```{r}
my_comparisons <- list(c("Pigs", "Rabbits"), c("Pigs", "Cows"), c("Human", "Cows"), c("Human", "Monkey") )
```

Then, we plot a [violin](https://mode.com/blog/violin-plot-examples/) plot using the function **ggviolin**.

```{r}
ggviolin(data = metadata.filtered.abricate.2, x = "Source.group", y = "NUM_FOUND", fill = "NUM_FOUND", color="Source.group", add = "boxplot", add.params = list(fill = "white"))+
  stat_compare_means(comparisons = my_comparisons, label = "p.signif")+ # Add significance levels
  stat_compare_means(label.y = 25) +                                      # Add global the p-value 
  theme(legend.position = 'none') +
  theme(axis.text.x=element_text(angle=30, hjust=1))
```

We see significant differences among the groups compared and global significant differences, but what are the specific p-values?

We use a t-test to see the P-values from pairwise comparisons.

```{r}
ggviolin(data = metadata.filtered.abricate.2, x = "Source.group", y = "NUM_FOUND", fill = "NUM_FOUND", color="Source.group", add = "boxplot", add.params = list(fill = "white"))+
  stat_compare_means(comparisons = my_comparisons, method = "t.test", aes(label=..p.adj..))+ # Add significance levels
  stat_compare_means(label.y = 25) +                                      # Add global the p-value 
  theme(legend.position = 'none') +
  theme(axis.text.x=element_text(angle=30, hjust=1))
```

We see that all p-values are lower than 0.05, suggesting significant (statistical) differences exist in the number of AMR genes found among the groups compared.


## Finally, we plot some AMR genes on our **phylogeny** to see their distribution

We use a similar approach as previously. As I mentioned previously, we want to set the root at the mid-point for our tree in most cases, normally because it makes it easier to read, and also because it displays the genetic relationships more accurately.

```{r}
new_aureus_tree <- midpoint.root(aureus_tree)
```

Here we plot [mecA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC139644/pdf/0335.pdf) on our tree, because it is one of the most important antimicrobial resistance genes found in S. aureus.

```{r}
gg1 <- ggtree(new_aureus_tree, layout = "rectangular", ladderize = T) 
gg1 <- gg1 %<+% metadata.filtered.abricate.2 + xlim(NA, 0.4) +
  geom_tippoint(aes(color = factor(Source.group)), size=1) 
gg1 <- gg1 + geom_tiplab(aes(fill = factor(mecA), palette="Pastel2"), color = "black", geom = "label", label.padding = unit(0.15, "lines"), label.size = 0, size=0.2, align=TRUE, family='mono', linetype = "dotted", linesize = .2)
#geom_treescale(x=0.08, y=35) #+
gg1 <- gg1 + theme(legend.position="right") + ggtitle("Staphylococcus aureus 571 genomes") +
  theme(plot.title = element_text(size = 12, hjust = 0.1, face = "bold")) +
  scale_colour_discrete(na.translate = F, name="Host_group") 
gg1
```

We see that mecA is found mostly in humans and cows.You can try to plot different AMR genes on your tree and see which ones might be more common in your host group (You can also find out in the paper).
