---
pdf_document: default
author: "Teddy Garcia-Aroca"
date: 'Last update: July 27th, 2022'
output:
  html_document: rmdformats::downcute
  pdf_document: default
title: "Project 2 | Andam Lab | Plotting Markov Jumps"
---

# Plotting results from **StarBeast** (Markov Jumps analyses).

**Note: this is just an example, since an approximation had to be used in BEAST to obtain these results**

In this tutorial, we will be showing examples of how to visualize results from Markov Jumps analyses from [StarBEAST](https://taming-the-beast.org/tutorials/StarBeast-Tutorial/), which can be used for a wide range of phylogenetic analyses and evolutionary data.

## Load packages, metadata, and tree as we did previously:
  
```{r results='hide', message=FALSE, warning=FALSE}
packages <- c("ape", "devtools", "dplyr", "plyr", "ggplot2", "readr", "ggpubr","ggnewscale",
              "rcompanion", "tidyverse", "tidyr", "ggsignif", "reshape", "ggtree", "phytools", "reshape2", "gridExtra")

# install packages that haven't been installed
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}

# load packages
invisible(lapply(packages, library, character.only = T))

#set working directory
setwd("/Volumes/bioinformaticslab/TGarciaAroca/Project2/RStudio/")

# Load metadata
metadata <- read.csv("/Volumes/bioinformaticslab/TGarciaAroca/Project2/metadata/metadata.csv", sep=",", header=T)
length(metadata$Number)

#Load tree
tree_file <- "/Volumes/bioinformaticslab/TGarciaAroca/Project2/datasets/core_genome/raxml/RAxML_bestTree.aureus_raxml_out"
aureus_tree <- read.tree(tree_file)
length(aureus_tree$tip.label)

#Filter metadata
metadata.filtered <- metadata[metadata$ERR.number.tips.filtered %in% aureus_tree$tip.label,] #Extract 
length(metadata.filtered$ERR.number.tips.filtered)

```

## Now, let's load results from **StarBeast**.

The results can be found in the folder called "markov_jumps_example" in the github repository (datasets/core_genome/markov_jumps_example/). These results have been adjusted to fit the tree and metadata columns.

```{r}
#Filter metadata
markov.jumps <- read.csv("/Volumes/bioinformaticslab/TGarciaAroca/Project2/datasets/core_genome/markov_jumps_example/aureus_markov_jumps.example.data.csv")
length(markov.jumps$ERR.number.tips.filtered)
```

### Merge the two datasets (metadata and **starBeast** results)

We merge them as we did in previous examples (**Note: since the column names are the same for both datasets, we merge then by ERR.number.tips.filtered**)

```{r}
metadata.filtered.markov <- merge(metadata.filtered, markov.jumps, by="ERR.number.tips.filtered")
length(metadata.filtered.markov$ERR.number.tips.filtered)
```

We see that we only have 571 datapoints as we did previously. However, this is a dataset with many columns. Let's just select the columns we need for this part:

```{r}
metadata.filtered.markov.subset <- select(metadata.filtered.markov, ERR.number.tips.filtered, Source.group, Human, Pigs, Cows, Goats, Birds, Rabbits,Sheep, Horses, Rodents, Carnivora,Bat, Marine, Unknown, Unclassified)
head(metadata.filtered.markov.subset, n=3)
```

Now we have a data frame that is easier to work with.

## Plot markov jumps by host

In order to do that, we have to first make a "long" dataset, to be able to better represent our data in boxplots (since we have many columns). For this we used a function called "pivot_longer" from the **tidyr** package.

```{r}
metadata.filtered.markov.subset.long <- pivot_longer(metadata.filtered.markov.subset, cols=3:16, names_to = "Host_group", values_to = "jumps")
length(metadata.filtered.markov.subset.long$ERR.number.tips.filtered)
head(metadata.filtered.markov.subset.long, n=3)
```

We can do the same thing with **reshape2** and this approach was most commonly used until it was replaced recently with the **tidyr** approach described above.

```{r}
metadata.filtered.markov.subset.melt <-melt(metadata.filtered.markov.subset, id=c("ERR.number.tips.filtered","Source.group"))
length(metadata.filtered.markov.subset.melt$ERR.number.tips.filtered)

#And fix the column names
colnames(metadata.filtered.markov.subset.melt) <- c("ERR.number.tips.filtered","Source.group", "Host_group", "jumps")
```

As you see, this is quite a long dataset with **7994** datapoints.

Now we plot like we did previously.

```{r}
metadata.filtered.markov.subset.melt$Host_group <- with(metadata.filtered.markov.subset.melt, reorder(Host_group, -jumps))
	ggplot(metadata.filtered.markov.subset.melt, aes(x =Host_group, y = jumps, fill=Host_group)) + 
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
   labs(
    title = "Comparisons number of host jumps by host groups",
    x = "Host groups",
    y = "Number of jumps"
    )
```

## Make a boxplot instead

Since this is numerical data, we can make a boxplot to at the variation across host groups.

```{r}
metadata.filtered.markov.subset.melt$Host_group <- with(metadata.filtered.markov.subset.melt, reorder(Host_group, -jumps))
	ggplot(metadata.filtered.markov.subset.melt, aes(x =Host_group, y = jumps, fill=Host_group)) + 
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
   labs(
    title = "Comparisons number of host jumps by host groups",
    x = "Host groups",
    y = "Number of jumps"
    )
```

As you see, this is way more informative. Host jumps from humans are higher, but there is also more variation in the amount of jumps.

## Plot violin plots

Let's compare some of those groups like we did previusly, but this time, let's leave humans out. We want to first create an object with the **pairwise comparisons** we want to make:

```{r}
my_comparisons <- list(c("Cows", "Rabbits"), c("Sheep", "Cows"), c("Sheep", "Rabbits"), c("Sheep", "Marine") )
```

## Now subset the data

```{r}
rabbits.cows.sheep.marine.jumps <-subset(metadata.filtered.markov.subset.melt, Host_group==c("Cows", "Rabbits","Sheep", "Marine"))
```

Then, we plot a [violin](https://mode.com/blog/violin-plot-examples/) plot using the function **ggviolin**.

```{r}
ggviolin(data = rabbits.cows.sheep.marine.jumps, x = "Host_group", y = "jumps", fill = "Host_group", color="Host_group", add = "boxplot", add.params = list(fill = "white"))+
  stat_compare_means(comparisons = my_comparisons, label = "p.signif")+ # Add significance levels
  stat_compare_means(label.y = 7) +                                      # Add global the p-value 
  theme(legend.position = 'none') +
  theme(axis.text.x=element_text(angle=30, hjust=1))
```

We see significant differences among the groups compared and global significant differences, but what are the specific p-values?

You can try to do what we did last time to see the especific P-values from pairwise comparisons.

## Finally, want to plot these jumps on our **phylogeny** to see their distribution

We use a similar approach as previously. As I mentioned previously, we want to set the root at the mid-point for our tree in most cases, normally because it makes it easier to read, and also because it displays the genetic relationships more accurately.

```{r}
new_aureus_tree <- midpoint.root(aureus_tree)
```

Here we plot the values for Cows, Rabbits, and Sheep that we were using previously.

First, let's extract those values from the original dataset.

```{r}
metadata.filtered.markov.for.tree <- metadata.filtered.markov[, c("ERR.number.tips.filtered", "Source.group", "Cows", "Rabbits", "Sheep")]
```

**IMPORTANT**: this was not demonstrated previously (sorting dataset to match tree label order).

Sort the dataset so that it follows the same order of tips as the tree:

```{r}
metadata.filtered.markov.for.tree.sorted <- metadata.filtered.markov.for.tree[match(new_aureus_tree$tip.label, metadata.filtered.markov.for.tree$ERR.number.tips.filtered),]
```

Then, we create a tree object.

```{r}
tree <- ggtree(new_aureus_tree, layout = "rectangular", ladderize = T) 
tree <- tree %<+% metadata.filtered.markov.for.tree.sorted + xlim(NA, 0.8) +
  geom_tippoint(aes(color = factor(Source.group)), size=1) + 
  geom_tiplab(color = "black", geom = "label", 
               label.padding = unit(0.15, "lines"), label.size = 0, 
               size=0.8, align=TRUE, family='mono', linetype = "dotted", linesize = .2)
```

Now we add the layers we want:

```{r}
cows <- select(metadata.filtered.markov.for.tree.sorted, Cows)
rownames(cows) <- new_aureus_tree$tip.label

tree1 <- gheatmap(tree, cows, offset=0.1, width=.05,
                       colnames_angle=60) +
           scale_fill_viridis_c(option="A", name="Cow jumps")
tree1
```

Now, we add rabbits:

```{r}
rabbits <- select(metadata.filtered.markov.for.tree.sorted, Rabbits)
rownames(rabbits) <- new_aureus_tree$tip.label

tree2 <- tree1 + new_scale_fill()

tree2 <- gheatmap(tree1, rabbits, offset=0.15, width=.05,
                       colnames_angle=60) +
           scale_fill_viridis_c(option="D", name="Rabbit jumps")
tree2
```

Feel free to try adding the other layers.